apiserver/clusterservice/scaleimpl.go

```diff
@@ -292,49 +334,179 @@ func ScaleDown(deleteData bool, clusterName, replicaName, ns string) msgs.ScaleD
+		response.Status.Msg = "Nothing to scale, the cluster is currently " +
+			"shutdown"
+		return response
+	}
+
+	selector := fmt.Sprintf("%s=%s,%s=%s", config.LABEL_PG_CLUSTER, clusterName,
+		config.LABEL_PGHA_ROLE, "master")
+	primaryList, err := kubeapi.GetPods(apiserver.Clientset, selector, ns)
 	if err != nil {
+		log.Error(err)
 		response.Status.Code = msgs.Error
 		response.Status.Msg = err.Error()
 		return response
+	} else if len(primaryList.Items) == 0 {
+		response.Status.Code = msgs.Error
+		response.Status.Msg = "no primary found for this cluster"
+		return response
+	} else if len(primaryList.Items) > 1 {
+		response.Status.Code = msgs.Error
+		response.Status.Msg = "more than one primary found for this cluster"
+		return response
 	}
-	if len(replicaList.Items) == 0 {
+	primaryDeploymentName := primaryList.Items[0].Labels[config.LABEL_DEPLOYMENT_NAME]
+
+	// selector in the format "pg-cluster=<cluster-name>,pg-ha-scope=<cluster-name>"
+	// which will grab the primary and any/all replicas
+	selector = fmt.Sprintf("%s=%s,%s=%s", config.LABEL_PG_CLUSTER, clusterName,
+		config.LABEL_PGHA_ROLE, "replica")
+	replicaList, err := kubeapi.GetPods(apiserver.Clientset, selector, ns)
+	if err != nil {
 		response.Status.Code = msgs.Error
-		response.Status.Msg = "no replicas found for this cluster"
+		response.Status.Msg = err.Error()
 		return response
 	}
 
-	//validate the replica name that was passed
-	replica := crv1.Pgreplica{}
-	found, err := kubeapi.Getpgreplica(apiserver.RESTClient, &replica, replicaName, ns)
-	if !found || err != nil {
-		log.Error(err)
+	// Return an error if trying to scale down the primary but replicas still remain in the
+	// cluster.  Otherwise if no replicas remain then proceed with scaling the primary
+	// deployment to 0 and set the cluster to a 'shutdown' status.
+	if shutdown && len(replicaList.Items) > 0 {
 		response.Status.Code = msgs.Error
-		if !found {
-			response.Status.Msg = replicaName + " replica not found"
-		} else {
+		response.Status.Msg = fmt.Sprintf("Replicas still exist for cluster %s. All replicas "+
+			"must be removed prior shutting down the database cluster.", clusterName)
+		return response
+	} else if shutdown {
+
+		log.Debugf("no replicas remain for cluster %s, scaling primary deployment %s to 0",
+			clusterName, primaryDeploymentName)
+
+		clusterServices, err := scaleClusterDeployments(cluster, 0)
+		if err != nil {
+			response.Status.Code = msgs.Error
 			response.Status.Msg = err.Error()
+			return response
+		}
+
+		// set the status of the pgcluster to 'pgcluster Shutdown' with associated message
+		message := "Cluster shutdown by scaling down the cluster"
+		if err := kubeapi.PatchpgclusterStatus(apiserver.RESTClient, crv1.PgclusterStateShutdown,
+			message, &cluster, ns); err != nil {
+			response.Status.Code = msgs.Error
+			response.Status.Msg = err.Error()
+			return response
+		}
+
+		if err := publishClusterShutdown(&cluster); err != nil {
```

Comment:
Given this is occuringly synchronously, we do'nt need an event, but just something as part of the response to indicate to the user that a shutdown is occurring.

Reply:
Or if the actual event is the 1shutdown` itself, which does occur async, having that event sent from the Operator.

Reply:
Now that shutdowns occur asynchronously with a pgcluster update, the shutdown event remains but has been moved from the API server to the Operator.